### PR TITLE
Support unicast query and answer

### LINKED ISSUE #93: Querying does not work for cases when advertising is delegated to another host.
### Your environment.
  - Version: v0.0.5
  - Browser: N/A
  - Other Information - N/A

### What did you do?
I have a code that advertises A records on behalf of other hosts. Simple code to do that is (this code runs on a host with a different IP address):

```go
package main

import (
	"github.com/davecheney/mdns"
	"fmt"
	"time"
)

func main() {
	err := mdns.Publish("test.local 60 IN A 10.0.1.0")
	if err != nil {
		panic(err)
	}
	fmt.Println("ok")
	time.Sleep(60 * time.Minute)
}
```

Then query it using code as in examples:

```go
package main

import (
	"context"
	"fmt"
	"net"

	"github.com/pion/mdns"
	"golang.org/x/net/ipv4"
)

func main() {
	addr, err := net.ResolveUDPAddr("udp", mdns.DefaultAddress)
	if err != nil {
		panic(err)
	}

	l, err := net.ListenUDP("udp4", addr)
	if err != nil {
		panic(err)
	}

	server, err := mdns.Server(ipv4.NewPacketConn(l), &mdns.Config{})
	if err != nil {
		panic(err)
	}
	answer, src, err := server.Query(context.TODO(), "test.local")
	if err != nil {
		panic(err)
	}
	fmt.Println(answer)
	fmt.Println(src)
	fmt.Println(err)
}
```

### What did you expect?

Querying using this library should work. It works using avahi client and it works using macos dns-sd. I expect that while the above code is running, querying using this library for "test.local" returns correct result.

### What happened?

It hangs indefinitely.

### PR DESCRIPTION
**Pull request by edaniels**
- Addresses #155
- Fixes #93

This also ends up internally supporting a good chunk of IPv6

### CODE REVIEW

**Comment by edaniels** [comment]
Blocked on https://github.com/pion/mdns/pull/173

File: conn.go [hunk 1]
```diff
@@ -556,6 +781,56 @@ func (c *Conn) start(inboundBufferSize int, config *Config) { //nolint gocognit
...
+			defer func() {
+				readerEnded <- struct{}{}
+			}()
+			readerStarted <- struct{}{}
+			c.readLoop("uni6", c.unicastPktConnV6, inboundBufferSize, config)
+		}()
+	}
+	for i := 0; i < numReaders; i++ {
+		<-readerStarted
+	}
```
Comment (SimonVerkada) [inline conn.go:827]: Could achieve something similar with `sync.waitGroup`

**Review by SimonVerkada — APPROVED** [review]
Looks good to me, small comments to consider

File: conn.go (hunk 1 above)
Comment (edaniels) [reply conn.go:827]: so I was thinking that but didn't feel like bothering with the race condition panics around waitgroups

File: conn.go [hunk 2]
```diff
@@ -390,42 +542,108 @@ func (c *Conn) sendAnswer(name string, ifIndex int, addr net.IP) {
...
+}
+
+func (c *Conn) readLoop(name string, pktConn ipPacketConn, inboundBufferSize int, config *Config) { //nolint:gocognit
 	b := make([]byte, inboundBufferSize)
 	p := dnsmessage.Parser{}
 
 	for {
-		n, cm, src, err := c.socket.ReadFrom(b)
+		_ = b
+		b := make([]byte, 1500)
```
[the remark this answers is not in the document]
Comment (edaniels) [reply conn.go:626]: old bad code!

**Comment by SimonVerkada** [comment]
Also, should consider adding additional unit tests

File: conn.go [hunk 3]
```diff
@@ -161,10 +243,22 @@ func (c *Conn) Close() error {
...
+		if err := c.unicastPktConnV4.Close(); err != nil {
+			return err
+		}
+	}
+
+	if c.unicastPktConnV6 != nil {
+		if err := c.unicastPktConnV6.Close(); err != nil {
+			return err
+		}
+	}
```
[the remark this answers is not in the document]
Comment (edaniels) [reply conn.go:260]: yeah will join them

**Comment by edaniels** [comment]
I'd love to add more unit tests, but I'm not really sure how to do that succinctly without getting very in the weeds with capturing the DNS packets or injecting in the unicast sockets. I'm open to suggestions for future PRs though!

**Pull request merged**